smithed_libraries/plugins/versioning/load.py

```diff
@@ -0,0 +1,237 @@
+        last_step_function = enumerate_step(
+            generate_enumerate,
+            opts.scoreholder,
+            set_version_path,
+            step,
+            value,
+            last_step_function,
+        )
+
+    enumerate_function = generate_function(
+        "enumerate",
+        Function(
+            f"scoreboard players reset {opts.scoreholder}.set load.status\n"
+            + "".join(
+                f"scoreboard players add {opts.scoreholder}.{name} load.status 0\n"
+                for name, _ in opts.version.named_parts()
+            )
+            + f"function {last_step_function}"
+        ),
+    )
+
+    return generate_tag(
+        "enumerate", merge=FunctionTag({"values": [enumerate_function]})
+    )
+
+
+def enumerate_step(
+    generate: Generator,
+    scoreholder: str,
+    set_version_path: str,
+    step: str,
+    value: int,
+    last_step_path: str | None,
+):
+    """A single step in the enumerate function
+
+    A step cooresponds to a version part (major, minor, etc).
+    """
+
+    step_body = (
+        "execute "
+        f"if score {scoreholder}.{step} load.status matches ..{value} "
+        f"unless score {scoreholder}.{step} load.status matches {value} "
+        f"run function {set_version_path}"
+    )
+
+    if last_step_path is not None:
+        step_body += (
+            "\n"
+            "execute "
+            f"unless score {scoreholder}.set load.status matches 1 "
+            f"if score {scoreholder}.{step} load.status matches ..{value} "
+            f"if score {scoreholder}.{step} load.status matches {value} "
+            f"run function {last_step_path}"
+        )
+
+    return generate(step, Function(step_body))
+
+
+def generate_load(ctx: Context, opts: VersioningOptions):
+    """This is the main versioning plugin
+
+    There are 3 parts to integrating lantern load for library support:
+     - Basic load tags
+        - Sets the step of the lantern load process to run everything else
+     - Enumeration
+        - Runs through every single pack of different versions
+        - Sets the highest version amongst them all
+     - Resolution
+        - Uses highest version to run the correct load file
+        - Disables the rest
+    """
+
+    with ctx.override(generate_namespace=opts.namespace, generate_prefix=""):
+        dependencies = ctx.generate[opts.lantern_load.tag_path].path("dependencies")
+
+        # Enumeration starts on the main version part
+        #  Then it checks sub-parts if the current part is the same
+        enumerate = enumerate_func(ctx, opts)
+
+        # Resolving involves using the highest version and booting off
+        #  cooresponding resources while killing off others
+        # Advancements get patched to "disable" if version check is **wrong**
+        resolve = resolve_func(ctx, opts)
+        resolve_advancements(ctx, opts)
+
+        load_tags(ctx, opts, dependencies, enumerate, resolve)
```

Comment:
I'm thinking that this file has a lot of logic that could be made more readable with bolt. It's fine as is but if you want to try it out you can use `subproject` to spawn a one-off compilation inside the plugin.

```py
config = {
    "require": ["bolt"],
    "data_pack": {"load": {"data/load/modules/main.bolt": f"@{__package__}/load.bolt"}},
    "pipeline": ["mecha"],
    "meta": {"bolt": {"entrypoint": "load:main"}},
}
ctx.require(subproject(config))
```

Anything emitted by bolt will get merged into the parent context at the end of the subpipeline.

Reply:
oh my, yea this would simplify things a ton. i'm debating whether it's worth to do now though. Also, `.bolt` code wouldn't be the best to read on Github.

I'm wondering though, what do u think of the use of `ctx.generate` here, a ton of the config allows me to very much control how versioning is applied here which allowed me to make a plugin which could apply to both gm4 and smithed which hopefully is a ton of the versioning spectrum.

I'm hoping that we can develop a `beet` or perhaps `bolt` `contrib` plugin with these sort of features, perhaps we can apply this clean up there (reminds me of our `beet-dependencies` ideas from a while back)!